### PR TITLE
make msbuildLogger required in its overload.

### LINKED ISSUE #43437: Addition of ILogger-accepting method in MSBuildWorkspace is a source breaking change
1. Have some tool that used MSBuildLocator and then MSBuildWorkspace.OpenSolutionAsync from the same method.
2. Move to a newer version of Roslyn that contains https://github.com/dotnet/roslyn/pull/42319

**Expected:** your code still works
**Actual:** it fails at runtime potentially with some strange failures.

One failure mode I'm seeing is the method that calls MSBuildWorkspace.OpenSolutionAsync triggers a load of ILogger from Microsoft.Build.Framework since that's now the method being chosen by overload resolution. If you are using MSBuildLocator in that same caller, that means it's loaded before your use of MSBuildLocator runs, causing MSBuildLocator to fail or end up with mismatched binaries.

### PR DESCRIPTION
**Pull request by jmarolf**
fixes #43437

### CODE REVIEW

File: src/Workspaces/Core/MSBuild/GlobalSuppressions.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+// See the LICENSE file in the project root for more information.
+
+// This file is used by Code Analysis to maintain SuppressMessage
+// attributes that are applied to this project.
+// Project-level suppressions either have no target or are given
+// a specific target and scoped to a namespace, type, member, etc.
+
+using System.Diagnostics.CodeAnalysis;
+
+[assembly: SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "Special case to avoid ILogger type getting loaded in downstream clients", Scope = "member", Target = "~M:Microsoft.CodeAnalysis.MSBuild.MSBuildWorkspace.OpenSolutionAsync(System.String,System.IProgress{Microsoft.CodeAnalysis.MSBuild.ProjectLoadProgress},System.Threading.CancellationToken)~System.Threading.Tasks.Task{Microsoft.CodeAnalysis.Solution}")]
```
Comment (sharwell) [inline src/Workspaces/Core/MSBuild/GlobalSuppressions.cs:12]: 💡 Move all these suppressions to the method(s) they apply to

File: src/Workspaces/Core/MSBuild/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -1,6 +1,6 @@
...
-Microsoft.CodeAnalysis.MSBuild.MSBuildWorkspace.OpenSolutionAsync(string solutionFilePath, System.IProgress<Microsoft.CodeAnalysis.MSBuild.ProjectLoadProgress> progress = null, Microsoft.Build.Framework.ILogger msbuildLogger = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> System.Threading.Tasks.Task<Microsoft.CodeAnalysis.Solution>
+Microsoft.CodeAnalysis.MSBuild.MSBuildWorkspace.OpenProjectAsync(string projectFilePath, Microsoft.Build.Framework.ILogger msbuildLogger, System.IProgress<Microsoft.CodeAnalysis.MSBuild.ProjectLoadProgress> progress = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> System.Threading.Tasks.Task<Microsoft.CodeAnalysis.Project>
+Microsoft.CodeAnalysis.MSBuild.MSBuildWorkspace.OpenProjectAsync(string projectFilePath, System.IProgress<Microsoft.CodeAnalysis.MSBuild.ProjectLoadProgress> progress = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> System.Threading.Tasks.Task<Microsoft.CodeAnalysis.Project>
```
Comment (sharwell) [inline src/Workspaces/Core/MSBuild/PublicAPI.Unshipped.txt:4]: 💭 Was this not already in the shipped API file?

File: src/Workspaces/Core/MSBuild/GlobalSuppressions.cs (hunk 1 above)
Comment (jmarolf) [reply src/Workspaces/Core/MSBuild/GlobalSuppressions.cs:12]: Is that the official policy for the IDE team? I typically only put suppression messages in the code file if we indent to remove them one day. As is stands this situation is expected to be permanent, so what is the advantage of polluting the code file with these suppressions? They are not something future maintainers are expected to reason about correct?

File: src/Workspaces/Core/MSBuild/PublicAPI.Unshipped.txt (hunk 2 above)
Comment (jmarolf) [reply src/Workspaces/Core/MSBuild/PublicAPI.Unshipped.txt:4]: I do not know what the policy for moving from unshipped to shipped is but I do not believe these apis were ever shipped in a non-beta package

File: src/Workspaces/Core/MSBuild/GlobalSuppressions.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/MSBuild/GlobalSuppressions.cs:12]: Yeah we don't use GlobalSuppressions.cs. Just #pragma disable and comment that as needed.

File: src/Workspaces/Core/MSBuild/PublicAPI.Unshipped.txt (hunk 2 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/MSBuild/PublicAPI.Unshipped.txt:4]: Correct, these haven't shipped in a final form yet -- 16.6 ship would have been that first point.

File: src/Workspaces/Core/MSBuild/Microsoft.CodeAnalysis.Workspaces.MSBuild.csproj [hunk 3]
```diff
@@ -28,6 +28,7 @@
   </ItemGroup>
   <ItemGroup>
     <PackageReference Include="Microsoft.Build" Version="$(MicrosoftBuildVersion)" ExcludeAssets="Runtime" PrivateAssets="All" />
+    <PackageReference Include="Microsoft.Build.Framework" Version="[$(MicrosoftBuildVersion),)" ExcludeAssets="Runtime" />
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/MSBuild/Microsoft.CodeAnalysis.Workspaces.MSBuild.csproj:31]: @jmarolf how is this different than just writing `Version="$(MicrosoftBuildVersion)"`?

Comment (jmarolf) [reply src/Workspaces/Core/MSBuild/Microsoft.CodeAnalysis.Workspaces.MSBuild.csproj:31]: just tested this. You are right, they are equivalent

**Pull request merged**